src/rgw/driver/rados/rgw_data_sync.cc

```diff
@@ -4468,6 +4468,20 @@ class RGWBucketSyncSingleEntryCR : public RGWCoroutine {
             tn->log(0, "entry with empty obj name, skipping");
             goto done;
           }
+
+          // if object lock is enabled on either, the other should follow as well
+          if (sync_pipe.source_bucket_info.obj_lock_enabled() != sync_pipe.dest_bucket_info.obj_lock_enabled()) {
+            set_status("skipping entry due to object lock mismatch");
+            tn->log(0, SSTR("skipping entry due to object lock mismatch: " << key));
+            goto done;
+          }
+          // make sure versioned object only lands on versioned bucket
+          if (!key.instance.empty() && !sync_pipe.dest_bucket_info.versioned()) {
```

Comment:
do we know whether `key.instance.empty()` matches null versions? is this behavior the same for full- vs incremental sync?

i see `RGWBucketSyncSingleEntryCR` has a `bool versioned` that comes from `rgw_bi_log_entry::is_versioned()`, along with a `bool null_verid` from `rgw_bi_log_entry::is_null_verid()`

maybe those are relevant here, although `RGWBucketFullSyncCR` passes false for both

Reply:
> do we know whether key.instance.empty() matches null versions?

no it doesn't.

there was a discussion around those vars here: https://github.com/ceph/ceph/pull/62594
but what i could understand eventually was, `is_null_verid` is only implemented for delete (https://github.com/ceph/ceph/pull/62594#issuecomment-2769747476) and the same for `is_versioned()` for null version id.